integration/helpers/archive.go

```diff
@@ -0,0 +1,154 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package helpers
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"context"
+	"io"
+	"log/slog"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/trace"
+)
+
+// CompressDirToZipFile compresses a directory into `.zip` format,
+// preserving the relative file path structure of the source directory.
+func CompressDirToZipFile(ctx context.Context, sourcePath, destPath string) (err error) {
+	archive, err := os.Create(destPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		_ = archive.Close()
+		if err != nil {
+			if err := os.Remove(destPath); err != nil {
+				slog.ErrorContext(ctx, "failed to remove archive", "error", err)
+			}
+		}
+	}()
+
+	zipWriter := zip.NewWriter(archive)
+	err = filepath.Walk(sourcePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if info.IsDir() {
+			return nil
+		}
+		file, err := os.Open(path)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer file.Close()
+		zipFileWriter, err := zipWriter.Create(filepath.Base(path))
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if _, err = io.Copy(zipFileWriter, file); err != nil {
+			return trace.Wrap(err)
+		}
+		return trace.Wrap(file.Close())
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err = zipWriter.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// CompressDirToTarGzFile compresses a directory into .tar.gz format,
+// preserving the relative file path structure of the source directory.
+func CompressDirToTarGzFile(ctx context.Context, sourcePath, destPath string) (err error) {
+	archive, err := os.Create(destPath)
```

Comment:
This still needs an explicit close with an error check, since it's writing the file. You could do this on 93, instead of duplicating it, since you're already using the named-error-return pattern.